hinted handoff: handle hints on separate connection and scheduling group

piodul opened the pull request:
Introduce a new verb dedicated for receiving and sending hints: `HINT_MUTATION`. It is handled on the streaming connection, which is separate from the one used for handling mutations sent by coordinator during a write.

The intent of using a separate connection is to increase fairness while handling hints and user requests - this way, a situation can be avoided in which one type of requests saturate the connection, negatively impacting the other one.

Information about new RPC support is propagated through new gossip feature `HINTED_HANDOFF_SEPARATE_CONNECTION`.

Fixes #4974.

Tests: unit(release)

piodul commented:
@vladzcloudius

I tried testing this PR in order to observe if it gives any improvement. I tried the following test:

- Start 3 or 9 node ccm cluster,
- Set up following resources:

```
CREATE KEYSPACE keyspace1 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 2};
CREATE TABLE keyspace1.standard1 (
    key blob PRIMARY KEY,
    "C0" blob,
    "C1" blob,
    "C2" blob,
    "C3" blob,
    "C4" blob
);
```

- Optionally, created a materialized view like this: `CREATE MATERIALIZED VIEW IF NOT EXISTS keyspace1.mv AS SELECT key, "C0", "C1", "C2", "C3" FROM keyspace1.standard1 WHERE "C0" IS NOT NULL AND key > 0x PRIMARY KEY ("C0", key);`
- Start cassandra-stress with 50-2000 threads,
- Stop one node for 1 minute, then start again,
- Observe metrics in grafana.

I looked at all dashboards in grafana from `scylla-monitoring`, and additionally at hint sending and writing numbers, and did not see any noticeable differences between the versions. I tried different combinations of parameters (with/without the view, 3 or 9 nodes, number of threads and write rate for `cassandra-stress`).

The test was run locally, on a cluster which was setup with ccm.

I suspect that the test setup or implementation is wrong. To be frank, I'm also not sure what kind of differences in behavior to look for. Could you tell me how to set up the test correctly?

gleb-cloudius commented on message/messaging_service.cc:
```diff
@@ -484,6 +484,8 @@ static constexpr unsigned do_get_rpc_client_idx(messaging_verb verb) {
     case messaging_verb::MUTATION_FAILED:
         return 3;
+    case messaging_verb::HINT_MUTATION:
+        return 4;
```
Why not use streaming connection?

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -258,6 +258,41 @@ class shared_mutation : public mutation_holder {
 };
 
+// shared mutation, but gets sent as a hint
+class hint_mutation : public mutation_holder {
```
You can inherit from shared_mutation

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -258,6 +258,41 @@ class shared_mutation : public mutation_holder {
+    explicit hint_mutation(const mutation& m) : hint_mutation(frozen_mutation_and_schema{freeze(m), m.schema()}) {
+    }
+    virtual bool store_hint(db::hints::manager& hm, gms::inet_address ep, tracing::trace_state_ptr tr_state) override {
+            return hm.store_hint(ep, _schema, _mutation, tr_state);
```
It does not make send to store a hint for a hint.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -258,6 +259,31 @@ class shared_mutation : public mutation_holder {
+    explicit hint_mutation(frozen_mutation_and_schema&& fm_a_s) : shared_mutation(std::move(fm_a_s)){
+    }
+    explicit hint_mutation(const mutation& m) : shared_mutation(m) {
+    }
```
I think you can do `using shared_mutation::shared_mutation`

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -258,6 +259,31 @@ class shared_mutation : public mutation_holder {
+    }
+    virtual bool store_hint(db::hints::manager& hm, gms::inet_address ep, tracing::trace_state_ptr tr_state) override {
+        return false;
+    }
```
Why not exception? But I do not insist.

piodul replied:
I wouldn't like to risk it. I can see a possibility where `store_hint` could be called at shutdown (`drain_on_shutdown` -> `timeout_cb` -> `hint_to_dead_endpoints` -> `store_hint`).

gleb-cloudius replied:
You are calling send_to_endpoint with disabled hints, this should not happen

vladzcloudius replied:
I agree with @gleb-cloudius if this one is hit it would definitely be a bug and we would want to know about this ASAP. You may keep it as 'return false' but then, please, add a very scary error message there.

piodul replied:
Ok. In the end I changed it into an exception throw.

vladzcloudius replied:
The message in the exception is confusing. It's not that you would try to store a hint for a second time but rather you would store a hint for a hint.

piodul replied:
Fixed

vladzcloudius commented on message/messaging_service.cc:
```diff
@@ -1276,4 +1277,17 @@ future<> messaging_service::send_paxos_learn(msg_addr id, clock_type::time_point
+}
+future<> messaging_service::send_hint_mutation(msg_addr id, clock_type::time_point timeout, const frozen_mutation& fm, std::vector<inet_address> forward,
+        inet_address reply_to, unsigned shard, response_id_type response_id, std::optional<tracing::trace_info> trace_info) {
+    return send_message_oneway_timeout(this, timeout, messaging_verb::MUTATION, std::move(id), fm, std::move(forward),
```
Should not the verb be `messaging_verb::HINT_MUTATION`?

gleb-cloudius replied:
Ha. That's a good catch!

piodul replied:
Oops... Fixed!

vladzcloudius replied:
Hmmm... I still see `MUTATION` verb. Are you sure you pushed your changes?

piodul replied:
Well, I did, but I applied the fix to a wrong commit. It should be all in order now.

vladzcloudius commented on message/messaging_service.cc:
```diff
@@ -1276,4 +1277,17 @@ future<> messaging_service::send_paxos_learn(msg_addr id, clock_type::time_point
+future<> messaging_service::send_hint_mutation(msg_addr id, clock_type::time_point timeout, const frozen_mutation& fm, std::vector<inet_address> forward,
+        inet_address reply_to, unsigned shard, response_id_type response_id, std::optional<tracing::trace_info> trace_info) {
+    return send_message_oneway_timeout(this, timeout, messaging_verb::MUTATION, std::move(id), fm, std::move(forward),
+        std::move(reply_to), std::move(shard), std::move(response_id), std::move(trace_info));
```
Some nitpicks (I'm surprised @gleb-cloudius missed these! :)):
`std::move(shard)` seems to be pointless - it's type is `unsigned`.
The same about std::move(response_id) which is uint64_t. Although I would keep this in case it changes in the future like `inet_address` did.

gleb-cloudius replied:
I had to leave something to you!

piodul replied:
It seems that some other handlers have the same problem. I got rid of moves of primitive integral types (`unsigned`, `uint32_t` and `uint64_t` - left alone typedefs such as `response_id`).

vladzcloudius commented on service/storage_proxy.cc:
```diff
@@ -258,6 +259,28 @@ class shared_mutation : public mutation_holder {
+            tracing::trace_state_ptr tr_state) override {
+        tracing::trace(tr_state, "Sending a hint to /{}", ep);
+        auto& ms = netw::get_local_messaging_service();
+        return ms.send_hint_mutation(netw::messaging_service::msg_addr{ep, 0}, timeout, *_mutation,
```
Why `cpu_id` is hard coded to `0`?

vladzcloudius replied:
I don't see where you send this on a separate (or streaming) socket.
Do I miss something?

gleb-cloudius replied:
cpu_id is always ignored. It is fine to hardcode it to whatever.

piodul replied:
@vladzcloudius Relevant changes are situated in `messaging_service.cc`. The client for given request is chosen based on `s_rpc_client_idx_table` array, which is initialized based on `do_get_rpc_client_idx` function.

vladzcloudius replied:
> @vladzcloudius Relevant changes are situated in `messaging_service.cc`. The client for given request is chosen based on `s_rpc_client_idx_table` array, which is initialized based on `do_get_rpc_client_idx` function.

Oh, I see it now. Thanks.

vladzcloudius commented:
@piodul About the test: an indication of the problem is an increased latency of regular writes beyond what <main class shares>:<streaming shares> would define.

In order to make this easier I'd recommend to use larger payload than the default c-s one, e.g. 10KB (the default is ~300B).

You may also want to make the node to be down for a little longer than 1 minute in order to allow about 5-10m worth hints to get accumulated (here I refer the amount of time it would take a node to receive all of these hints).

Then, you may want to switch to CL=ALL when hints are being sent so that your query latency get hit by those hints for sure. 

(Note that you can't use CL=ALL during the phase when you generate hints).

During the hints sending phase the rate of queries should correspond to ~30%-70% per-shard CPU load without hints. 
However with hints being sent CPU load should be 100%.

So, the expected observation should be:
 * Without your patches - latency degradation should be beyond 20% compared to the same writes rate without any hints.
 * With your patches a latency degradation should be not more than 20%.

Please, let me know if you need further assistance with a test case. We can jump on a quick call if needed.

piodul commented on service/storage_proxy.cc:
```diff
@@ -2230,6 +2253,20 @@ future<> storage_proxy::send_to_endpoint(
+            std::move(target),
+            {},
+            db::write_type::SIMPLE,
+            _stats,
```
I changed the first call to use `_stats` instead, so that both use the same overload (the one with the actual logic). This way we don't need to introduce yet another overload.

vladzcloudius replied:
Fine with me.

piodul commented:
Fixed the spelling of the "move register_mutation handler to local lambda" commit.
Changed slightly the description of the last commit (getting rid of moves), as it actually gets rid of moves of `unsigned` values (there were no moves of `uint32_t`s and `uint64_t`s).

slivne commented:
@gleb-cloudius please review (hopefully bless) ... I want to try and  have sarna merge this in.

gleb-cloudius commented on message/messaging_service.cc:
```diff
@@ -994,7 +995,7 @@ void messaging_service::unregister_mutation() {
     inet_address reply_to, unsigned shard, response_id_type response_id, std::optional<tracing::trace_info> trace_info) {
     return send_message_oneway_timeout(this, timeout, messaging_verb::MUTATION, std::move(id), fm, std::move(forward),
-        std::move(reply_to), std::move(shard), std::move(response_id), std::move(trace_info));
+        std::move(reply_to), shard, std::move(response_id), std::move(trace_info));
```
For the future those cleanups should be in a separate patch.
